src/FreeAge/client/command_button.cpp

```diff
@@ -54,6 +54,32 @@ void CommandButton::SetAction(ActionType actionType, const Texture* texture, Qt:
   this->texture = texture;
 }
 
+CommandButton::State CommandButton::GetState(GameController* gameController) const {
+
+  if (GetType() == CommandButton::Type::ProduceUnit) {
+    if (!gameController->GetLatestKnownResourceAmount().CanAfford(
+            GetUnitCost(GetUnitProductionType()))) {
+      return State::CannotAfford;
+    }
+
+  } else if (GetType() == CommandButton::Type::ConstructBuilding) {
+    BuildingType buildingType = GetBuildingConstructionType();
+
+    // Check for the max limit
+    int max = GetBuildingMaxInstances(buildingType);
+    if (max != -1 && gameController->GetBuildingTypeCount(buildingType) >= max) {
+      return State::MaxLimitReached;
+    }
+
+    // Can afford must be the last check
```

Comment:
Maybe also state the reason for why it must be the last check, otherwise it might be unclear where this condition comes from